fairmq/shmem/Socket.h

```diff
@@ -125,23 +135,83 @@ class Socket final : public fair::mq::Socket
         return zmq::Connect(fSocket, address, fId);
     }
 
-    int64_t Send(MessagePtr& msg, int timeout = -1) override
+  private:
+    auto MakeMetaMsg(shmem::Message const& shmMsg) noexcept -> zmq_msg_t
     {
+        std::size_t const size = std::max(fMetadataMsgSize, sizeof(MetaHeader));
+        zmq_msg_t metaMsg;
+        auto rc = 0;
+
+        if (fMetadataMsgSize > size) {
+            // copy the MetaHeader into a larger buffer than needed
+            auto paddedMetaMsg = tools::MakePaddedUnique<MetaHeader>(
+                size, alignof(std::max_align_t), std::pmr::get_default_resource(), shmMsg.fMeta);
+
+            using Deleter = typename decltype(paddedMetaMsg)::deleter_type;
+            auto del = std::make_unique<Deleter>(paddedMetaMsg.get_deleter());
+
+            rc = zmq_msg_init_data(
+                &metaMsg,
+                paddedMetaMsg.release(),
+                size,
+                [](void* data, void* hint) {
+                    if (data) {
+                        assertm(hint, "hint exists");   // NOLINT
+                        // hint points to the deleter callable
+
+                        auto metaMsgDeleter = static_cast<Deleter*>(hint);
+                        (*metaMsgDeleter)(data);
+
+                        auto deleterDeleter = std::default_delete<Deleter>();
+                        deleterDeleter(metaMsgDeleter);
+                    }
+                },
+                // (lambda) const * -> void *
+                static_cast<void*>(const_cast<   // NOLINT
+                    std::add_pointer_t<std::remove_const_t<std::remove_pointer_t<Deleter>>>>(
+                        del.release()))
+                );
+            assertm(rc == 0, "msg init successful");   // NOLINT
+        } else {
+            rc = zmq_msg_init_size(&metaMsg, sizeof(MetaHeader));
+            assertm(rc == 0, "msg init successful");   // NOLINT
+            // copy the shmMsg::fMeta field into the zmq data buffer
+            auto data = zmq_msg_data(&metaMsg);
+            std::memcpy(data, &(shmMsg.fMeta), sizeof(MetaHeader));
+        }
+
+        return metaMsg;
+    }
+
+  public:
+    int64_t Send(mq::MessagePtr& msg, int timeout = -1) override
+    {
+        auto msgPtr = msg.get();
+        if (!msgPtr) {
+            return static_cast<int>(TransferCode::error);
+        }
+        assertm(dynamic_cast<shmem::Message*>(msgPtr), "given mq::Message is a shmem::Message");   // NOLINT
+        auto shmMsg = static_cast<shmem::Message*>(msgPtr);   // NOLINT(cppcoreguidelines-pro-type-static-cast-downcast)
+
+        auto metaMsg = MakeMetaMsg(*shmMsg);
+        // according to http://api.zeromq.org/master:zmq-msg-close we need to zmq_msg_close in all cases
+        // except on successfull send to not leak the msg.
+        auto closeMetaMsg = tools::CallOnDestruction([&metaMsg]() { zmq_msg_close(&metaMsg); });
```

Comment:
This is ugly. @rbx do you know if I can just always call `zmq_msg_close`?

On a second note, should I perhaps use [`shmem::ZMsg`](https://github.com/FairRootGroup/FairMQ/blob/73fd1b2c2a3c2784368aba2c7805b70c6ed9afdf/fairmq/shmem/Socket.h#L33) here?

Reply:
I understand the docs in the way that you can call zmg_msg_close when *you* no longer need the msg. It will still live on until it is transferred. So always closing is fine.

Reply:
done